src/main/java/org/traccar/broadcast/RedisBroadcastService.java

```diff
@@ -0,0 +1,206 @@
+import org.slf4j.LoggerFactory;
+import org.traccar.config.Config;
+import org.traccar.config.Keys;
+import org.traccar.model.BaseModel;
+import org.traccar.model.Device;
+import org.traccar.model.Event;
+import org.traccar.model.Permission;
+import org.traccar.model.Position;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import redis.clients.jedis.Jedis;
+import redis.clients.jedis.JedisPubSub;
+
+public class RedisBroadcastService implements BroadcastService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(RedisBroadcastService.class);
+
+    private final ObjectMapper objectMapper;
+
+    private final ExecutorService service = Executors.newSingleThreadExecutor();
+
+    private final Set<BroadcastInterface> listeners = new HashSet<>();
+
+    private final String url;
+    private final String pubsubChannel = "traccar:cast";
+
+    private final Jedis subscriberJedis;
+    private final Jedis publisherJedis;
+
+    private final String id;
+
+    public RedisBroadcastService(Config config, ObjectMapper objectMapper) throws IOException {
+        this.objectMapper = objectMapper;
+        url = config.getString(Keys.BROADCAST_ADDRESS);
+
+        subscriberJedis = new Jedis(url);
+        publisherJedis = new Jedis(url);
+
+        // id that will be used to identify this instance of the server
+        id = String.valueOf(System.currentTimeMillis());
+    }
+
+    @Override
+    public boolean singleInstance() {
+        return false;
+    }
+
+    @Override
+    public void registerListener(BroadcastInterface listener) {
+        listeners.add(listener);
+    }
+
+    @Override
+    public void updateDevice(boolean local, Device device) {
+        BroadcastMessage message = new BroadcastMessage();
+        message.setDevice(device);
+        sendMessage(message);
+    }
+
+    @Override
+    public void updatePosition(boolean local, Position position) {
+        BroadcastMessage message = new BroadcastMessage();
+        message.setPosition(position);
+        sendMessage(message);
+    }
+
+    @Override
+    public void updateEvent(boolean local, long userId, Event event) {
+        BroadcastMessage message = new BroadcastMessage();
+        message.setUserId(userId);
+        message.setEvent(event);
+        sendMessage(message);
+    }
+
+    @Override
+    public void updateCommand(boolean local, long deviceId) {
+        BroadcastMessage message = new BroadcastMessage();
+        message.setCommandDeviceId(deviceId);
+        sendMessage(message);
+    }
+
+    @Override
+    public void invalidateObject(boolean local, Class<? extends BaseModel> clazz, long id) {
+        BroadcastMessage message = new BroadcastMessage();
+        message.setChanges(Map.of(Permission.getKey(clazz), id));
+        sendMessage(message);
+    }
```

Comment:
There's a lot of duplication. Can we make a base class with all common stuff that doesn't change between implementations?

Reply:
Could be moved to BroadcastService?

Reply:
It's not needed for Noop implementation, so I would just make another base class like `BaseBroadcastService`.